core/src/main/java/io/ably/lib/util/ObjectCopyUtil.java

```diff
@@ -0,0 +1,98 @@
+package io.ably.lib.util;
+
+import java.util.Arrays;
+import java.util.HashMap;
+
+import io.ably.lib.rest.Auth;
+import io.ably.lib.types.ClientOptions;
+import io.ably.lib.types.ProxyOptions;
+
+public class ObjectCopyUtil {
```

Comment:
As far as I can see no. As you have more experience can you tell me if there is something else worth cloning?

Reply:
I'm afraid that I don't know :thinking: @QuintinWillison maybe you recall some other things that we should copy?

Reply:
My honest answer is that I don't know. However I think it's perhaps most useful to consider the isolation and protection we're looking to introduce:

- Objects which are not designed to be immutable (e.g. `ClientOptions`, regrettably) and which are passed by reference to an SDK API with the intention being that they are considered static from that point (e.g. `ClientOptions`) should have a snapshot taken by the SDK implementation, so that the SDK has its own deep copy going forwards. I mention deep copy, crucially, because we need to ensure arrays, maps, etc also cannot be changed later.
- Objects which are passed back from the SDK via a listener or callback, which are implemented with some kind of mutable primitive (e.g. `HashMap` or `ArrayList`), should be free to mutate by the app consuming them without then changing expectations or state of the SDK implementation in surprising ways

We're clearly somewhat hamstrung by the fact that many of our 'properties' on our classes suffer from:

- being settable when they should be read only
- are probably implemented as fields, rather than get and/or set methods

Changing these things, including the net conclusion that we should be using interfaces rather than classes in many cases, is realistically out-of-scope for our v`2` `major` bump milestone. So, for now, we're stuck with finding any quick wins that don't change API surface area.

Reply:
I am wondering if we can move this copy method to the ClientOptions directly and remove this util class? It should not be hard to maintain it later.

Reply:
Is it worth to keep scope of this PR of changing `ClientOptions` to move this PR forward @KacperKluka ?

Reply:
Yes, that's a good idea :+1: However, please change this PRs name to something like "Increase ClientOptions mutability safety by cloning objects" and make sure that the https://github.com/ably/ably-java/issues/771 is not yet resolved after merging this PR (as there might be other objects we should clone) :wink:

Reply:
@KacperKluka @ikbalkaya Here it is https://github.com/ably/ably-java/pull/827/commits/afbaf74b1ee3e9906f3203676c2d9c7ec0694062